Fix deploy to use valid docker image names including tags

BurtonR opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
This resolves the issue #19 that throws an error when using
a tag on the image name.

There are several variations to a valid docker image name.
Repository may, or may not, be included. Default is docker.io.
Username may, or may not, be included.
Repository name is always included.
An image tag may, or may not, be included. Default is 'latest'.


This update takes all those scenarios into account when getting
the authentication credentials to allow private functions
to be deployed from docker hub, or another repository

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
- [X] I have raised an issue to propose this change ([required](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md))
Resolves #19 

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
Updated unit tests in the `registryauth_test.go` file to include test cases without a repository and test cases with an image tag.

Also, verified by: 
* Create a private function (on hub.docker.com)
* `faas deploy -a`  with image as: `burtonr/pvt-tester`
* `faas deploy -a` with image as: `burtonr/pvt-tester:latest`
* `faas deploy -a` with image as: `burtonr/pvt-tester:0.1`
* `faas deploy` with a new template. Image as `tester`
* `faas deploy -a` with image as `burtonr.azurecr.io/burtonr/pvt-tester`
* `faas deploy -a` with image as `burtonr.azurecr.io/burtonr/pvt-tester:latest`

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [X] I've read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [X] I have signed-off my commits with `git commit -s`
- [X] I have added tests to cover my changes.
- [X] All new and existing tests passed.

alexellis commented on handlers/deploy.go:
```diff
@@ -190,8 +190,11 @@ func buildEnv(envProcess string, envVars map[string]string) []string {
 // BuildEncodedAuthConfig for private registry
 func BuildEncodedAuthConfig(basicAuthB64 string, dockerImage string) (string, error) {
 	// extract registry server address
+	if len(strings.Split(dockerImage, "/")) < 3 {
```
I think we can document this better.. when we read it again in 6 months it will look like magic

BurtonR replied:
Good point!

alexellis commented on test/registryauth_test.go:
```diff
@@ -47,7 +54,7 @@ func testValidEncodedAuthConfig(t *testing.T, user, password, imageName, expecte
 
 	if user != authConfig.Username {
-		t.Log("Auth config username mismatch", user, authConfig.Username)
+		t.Logf("Auth config username mismatch want %s, got: %s", user, authConfig.Username)
```
Great way to improve the readability of this test. Do tests fail if you keep the test code but revert the code fix?

BurtonR replied:
The tests were passing with this line in it's original state, but I was having trouble knowing what the problem was, so I fixed it :smile:

BurtonR replied:
I just re-read your comment. I misunderstood you the first time around...
Yes, by only adding in the additional test cases, the tests would fail. In the original state of the code, the repository _must_ be included, and no tag could be used. This is due to the assumptions that the `WithName` function parsed the input (from the docker library)

alexellis replied:
I was referring to the Logs usage / want/got values are really important in a test failure 👍

BurtonR commented:
I updated the description for the `BuildEncodedAuthConfig` function to _hopefully_ explain what was going on with the `strings.Split` to avoid any confusion later.

alexellis commented:
@BurtonR  / @rgee0 are we good to merge this?

BurtonR commented:
I'm confident this is ready to get merged

rgee0 commented on handlers/deploy.go:
```diff
@@ -187,11 +187,15 @@ func buildEnv(envProcess string, envVars map[string]string) []string {
 
-// BuildEncodedAuthConfig for private registry
+// BuildEncodedAuthConfig parses the image name for a repository, user name, and image name
+// If a repository is not included (ie: username/function-name), 'docker.io/' will be appended
```
Sorry to be pedantic but given comments are meant to help someone else understand then `appended` should be `prepended`

BurtonR replied:
That's a good point! I couldn't think of that word at the time! Good catch :smile:

BurtonR commented:
Updated the comment to use the appropriate terminology to avoid confusion. Thanks @rgee0!

alexellis commented:
OK I'll merge and cut a release. Please can you re-run your scenarios with the official image and then raise a PR to update faas-swarm in the faas repo? That way we can start benefiting from your hard work 💯 😄

alexellis commented:
@rgee0 thanks for the review 👍

alexellis commented on handlers/deploy.go:
```diff
@@ -187,11 +187,15 @@ func buildEnv(envProcess string, envVars map[string]string) []string {
 func BuildEncodedAuthConfig(basicAuthB64 string, dockerImage string) (string, error) {
-	// extract registry server address
+	// use docker.io if no repository was included
+	if len(strings.Split(dockerImage, "/")) < 3 {
```
I previously gave the feedback that we shouldn't use a magic number `3` - it may have been on the CLI repo.

What does 3 mean and how can we make sure it's clear going forward?

BurtonR replied:
My next PR in this repo will update that to be more clear.

|1|2|3|
|---|---|---|
|repository|username|image|

So, if there are less than 3, there's no repository included. During testing of this PR, I did also verify that this will still work as expected with _only_ an image name `function-1` for example

rgee0 replied:
Lets not over-think this.  Magic numbers should, as a minimum, be extracted into constants and given a meaningful name to assist readability.  `const fullImagePathLength int = 3` turns the existing code into `if len(strings.Split(dockerImage, "/")) < fullImagePathLength {` which even at a minimal level is more readable.

Taking it further we could wrap the left hand side into a fn called imagePathLength() we get:
```
if imagePathLength(dockerImage) < fullImagePathLength {
```

I lack imagination when naming things but I hope this helps demonstrate the point.

alexellis replied:
`fullImagePath` means nothing to a reader - the suggestion was to express this differently - via two named variables for 1 and 2 showing that 1 was username (namespace) and 2 was including a domain name.

rgee0 replied:
Apologies, didn't see the suggestion only the question.
